File: packages/algorithm/Leetcode/hot100/DP/最长递增子序列.ts
function lengthOfLIS(nums: number[]): number {
    // dp[i] 表示以 nums[i] 结尾的最长递增子序列的长度
    // 初始都为1，最差也有它自己
    const dp = new Array(nums.length + 1).fill(1);
    let maxLen = 1;

    for (let i = 0; i < nums.length; i++) {
        for (let j = 0; j < i; j++) {
            if (nums[j] < nums[i])
                dp[i] = Math.max(dp[i], dp[j] + 1);
        }
        maxLen = Math.max(maxLen, dp[i]);
    }

    return maxLen;
}

File: packages/program-design/DP/最长递增子序列.py
from typing import List


class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [1] * n
        
        ans = 0
        for i in range(n):
            for j in range(i, n):
                if nums[j] > nums[i]:
                    dp[j] = max(dp[j], dp[i] + 1)

            ans = max(dp[i], ans)

        return ans